### PR TITLE
Implement PWM for linux

### PR DESCRIPTION
**Pull request by hs0225**
- Implement PWM API for general linux.
- Add testsuite, and tested on rpi2
- rename enable into setEnable

IoT.js-DCO-1.0-Signed-off-by: Hosung Kim hs852.kim@samsung.com

### CODE REVIEW

**Comment by glistening** [comment]
Could you please use 2 spaces instead of 4 spaces for following iotjs coding style guide? (e.g. pwm.js)

File: src/module/iotjs_module_pwm.h [hunk 1]
```diff
@@ -41,10 +66,11 @@ class Pwm : public JObjectWrap {
   static Pwm* GetInstance();
   static const iotjs_jval_t* GetJPwm();
 
+  virtual int InitializePwmPath(PwmReqWrap* pwm_rea) = 0;
```
Comment (nova0821) [inline src/module/iotjs_module_pwm.h]: Maybe ``pwm_req``?

File: src/module/iotjs_module_pwm.h [hunk 2]
```diff
@@ -41,10 +66,11 @@ class Pwm : public JObjectWrap {
...
   virtual int Export(PwmReqWrap* pwm_req) = 0;
-  virtual int SetDutyCycle(PwmReqWrap* pwm_req) = 0;
   virtual int SetPeriod(PwmReqWrap* pwm_req) = 0;
-  virtual int Enable(PwmReqWrap* pwm_req) = 0;
+  virtual int SetDutyCycle(PwmReqWrap* pwm_req) = 0;
+  virtual int SetEnable(PwmReqWrap* pwm_req) = 0;
   virtual int Unexport(PwmReqWrap* pwm_rea) = 0;
```
Comment (nova0821) [inline src/module/iotjs_module_pwm.h]: ditto.

File: src/platform/arm-nuttx/iotjs_module_pwm-arm-nuttx.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ public:
+  explicit PwmArmNuttx(const iotjs_jval_t* jpwm);
+
+  static PwmArmNuttx* GetInstance();
+  virtual int InitializePwmPath(PwmReqWrap* pwm_req);
+  virtual int Export(PwmReqWrap* pwm_req);
+  virtual int SetPeriod(PwmReqWrap* pwm_req);
+  virtual int SetDutyCycle(PwmReqWrap* pwm_req);
+  virtual int SetEnable(PwmReqWrap* pwm_req);
+  virtual int Unexport(PwmReqWrap* pwm_rea);
```
Comment (nova0821) [inline src/platform/arm-nuttx/iotjs_module_pwm-arm-nuttx.cpp]: ditto.

File: src/js/pwm.js [hunk 4]
```diff
@@ -26,70 +60,107 @@ function PWM(pin, options, callback) {
...
+    var self = this;
     // Check arguments.
-    if (!util.isNumber(dutyCycle)) {
-        throw new TypeError('Bad arguments - dutyCycle');
+    if (!util.isNumber(period)) {
+        throw new TypeError('Bad arguments - period');
+    }
+
+    if (period < self._options.dutyCycle) {
+        console.log("%d %d", self._options.dutyCycle, period);
```
Comment (hs0225) [inline src/js/pwm.js]: It's my mistake. I fixed it. Thank you for review.

**Comment by hs0225** [comment]
@glistening Thanks, I fixed it.

File: src/module/iotjs_module_pwm.h (hunk 1 above)
Comment (hs0225) [inline src/module/iotjs_module_pwm.h]: I fixed it.

File: src/module/iotjs_module_pwm.h (hunk 2 above)
Comment (hs0225) [inline src/module/iotjs_module_pwm.h]: done.

File: src/platform/arm-nuttx/iotjs_module_pwm-arm-nuttx.cpp (hunk 3 above)
Comment (hs0225) [inline src/platform/arm-nuttx/iotjs_module_pwm-arm-nuttx.cpp]: done.

File: src/js/pwm.js [hunk 5]
```diff
@@ -16,82 +16,152 @@
...
-    }
+  self._options = {};
 
-    return pwm.export(this._pin, options, function(err) {
-        if (util.isFunction(callback)) {
-            callback(err);
-        }
-    });
-}
+  var pinPath = pin + ''; // make string
```
Comment (chokobole) [inline src/js/pwm.js]: how about  using `toString()`?

File: src/js/pwm.js [hunk 6]
```diff
@@ -16,82 +16,152 @@
...
+  return new PwmError(errno, operation, 'Unknown error');
+}
+
+
 // PWM(pin[, options][, callback])
 function PWM(pin, options, callback) {
-    if (!(this instanceof PWM)) {
-        return new PWM(pin, options, callback);
-    }
+  var self = this;
```
Comment (chokobole) [inline src/js/pwm.js]: is `self` needed? can we remove `self`?

File: src/js/pwm.js [hunk 7]
```diff
@@ -16,82 +16,152 @@
...
+}
 
-    return pwm.setDutyCycle(this._pin, period, function(err) {
-        if (util.isFunction(callback)) {
-            callback(err);
-        }
-    });
+// pwm.setPeriod(period[, callback])
+PWM.prototype.setPeriod = function (period, callback) {
+  var self = this;
```
Comment (chokobole) [inline src/js/pwm.js]: ditto

File: src/module/iotjs_module_pwm.cpp [hunk 8]
```diff
@@ -45,53 +45,140 @@ Pwm* Pwm::GetInstance() {
...
-  IOTJS_ASSERT(!"Not implemented");
+  JHANDLER_CHECK_ARGS(3, string, object, function);
+
+  // PwmReqWrap and string object(PwmReqData.device) are freed by AfterPwmWork.
+  PwmReqWrap* req_wrap = new PwmReqWrap(JHANDLER_GET_ARG(2, function));
+  PwmReqData* req_data = req_wrap->req();
+
+  const iotjs_jval_t* joption = JHANDLER_GET_ARG(1, object);
+  iotjs_jval_t jperiod = iotjs_jval_get_property(joption, "period");
+  iotjs_jval_t jduty_cycle = iotjs_jval_get_property(joption, "dutyCycle");
```
Comment (chokobole) [inline src/module/iotjs_module_pwm.cpp]: should destory `jperiod` and `jduty_cycle` in the last.

File: src/platform/arm-nuttx/iotjs_module_pwm-arm-nuttx.cpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+
+PwmArmNuttx::PwmArmNuttx(const iotjs_jval_t* jpwm)
+    : Pwm(jpwm) {
+}
+
+
+PwmArmNuttx* PwmArmNuttx::GetInstance()
+{
```
Comment (nova0821) [inline src/platform/arm-nuttx/iotjs_module_pwm-arm-nuttx.cpp]: Please check coding style here.

File: src/module/iotjs_module_pwm.cpp (hunk 8 above)
Comment (hs0225) [inline src/module/iotjs_module_pwm.cpp]: I missed it. Thank you for review.

**Comment by zherczeg** [comment]
LGTM

**Comment by yichoi** [comment]
LGTM

**Pull request merged**